web/src/app/components/timeline/timeline.component.ts

```diff
@@ -0,0 +1,400 @@
+import { AfterViewInit, Component, ElementRef, Input, ViewChild } from '@angular/core';
+import { HumanizeDurationLanguage, HumanizeDuration, HumanizeDurationOptions } from 'humanize-duration-ts';
+import * as d3 from 'd3';
+
+import { step189_2020 } from '../../../proto/step189_2020';
+import { ScaleLinear } from 'd3';
+
+/**
+ * Item holds all required data for one interval on the timeline.
+ */
+interface Item {
+  pushID: string;     // Push ID number
+  state: number;      // Final state of push
+  startTime: number;  // Start of push, in milliseconds
+  endTime: number;    // End of push, in milliseconds
+  row: number;        // Row number corresponds to y-position on timeine
+}
+
+@Component({
+  selector: 'app-timeline',
+  templateUrl: './timeline.component.html',
+  styleUrls: ['./timeline.component.scss']
+})
+
+export class TimelineComponent implements AfterViewInit {
+  /**
+   * Private data variables. 
+   * 
+   * Note that we use the non-null assertion operator ('!') to indicate to 
+   * the compiler that variables will always be declared and thus to not issue
+   * errors about possibilities of them being `null` or `undefined` since
+   * pushInfos and data are passed-in values.
+   */
+  @ViewChild('timeline') private timelineContainer!: ElementRef;
+  @Input() private pushInfos!: step189_2020.IPushInfo[] | null;
+  private data: Item[] = [];
+  private svg: any;
+  private x: ScaleLinear<number, number> = d3.scaleLinear();
+  private xAxis: Function = () => {};
+  private height: number = 0;
+  private width: number = 0;
+  private numRows: number = 0;
+
+  /**
+   * Constants.
+   */
+  private static readonly LANG_SERVICE: HumanizeDurationLanguage = new HumanizeDurationLanguage();
+  private static readonly HUMANIZER: HumanizeDuration = new HumanizeDuration(TimelineComponent.LANG_SERVICE);
+  private static readonly COLOR_LIGHT_GRAY: string = '#d3d3d3';
+  private static readonly MIN_INTERVAL_HEIGHT: number = 25;
+  private static readonly NSEC_PER_MSEC: number = 10 ** 6;
+  private static readonly STATE_TO_COLOR: { [index: number]: string } = {
+    5: '#34a853',
+    6: '#d50000',
+    10: '#2196f3',
+    12: '#d50000',
+    13: '#2196f3',
+    14: '#eee',
+    15: '#2196f3',
+    16: '#d50000',
+    17: '#eee',
+    18: '#34a853',
+    19: '#eee'
+  };
+
+  /**
+   * Draws the initial timeline and updates it on any change.
+   * @param changes Hashtable of changes
+   */
+  ngAfterViewInit(): void { this.createTimeline(); }
+
+  /**
+   * Extracts the pushID, state, and start and end time for each push
+   * in pushInfos and inserts them into Item interfaces, which
+   * are collectively stored in an array.
+   * @param pushInfos Array of pushes for one push def
+   */
+  private populateData(pushInfos: step189_2020.IPushInfo[] | null): void {
+    if (!pushInfos) { return; }
+
+    pushInfos.forEach(pushInfo => {
+      if (!pushInfo) { return; }
+      const pushes = pushInfo.stateInfo;
```

Comment:
Ying mentioned this on mine but "pushes" is describing a single push and it is actually "states"